src/core/thread/mesh_forwarder.hpp

```diff
@@ -420,6 +420,21 @@ class MeshForwarder : public InstanceLocator, private NonCopyable
         kAnycastService,
     };
 
+    /**
+     * This function pointer can be used to process messages dropped due to age.
+     *
+     * @param[in]  aMessage     A pointer to the dropped message.
+     * @param[in]  aContext     A pointer to callback context.
+     *
+     */
+    typedef void (*AgedMessageDroppedCallback)(Message *aMessage, void *aContext);
+
+    typedef struct
+    {
+        Message *mMessage;
+        bool    *mWasDropped;
+    } AgedMessageDroppedContext;
```

Comment:
The callback model to detect if message is freed can work, but I think we may be able to address this with simpler solution. We can use `SetDoNotEvict()` flag in `Message` to disallow it being removed from `RemoveAgeMessages()`. I can submit a PR with this.

Reply:
Submitted https://github.com/openthread/openthread/pull/9348 as a different way to address this.
Thoughts?

Reply:
I was not aware of `SetDoNotEvict()`. I like your solution better, it is simpler and avoids boilerplate.
I am declining this PR.
Thank you!